packages/styletron-engine-monolithic/src/inject-style-prefixed.js

```diff
@@ -0,0 +1,98 @@
+// @flow
+
+declare var __DEV__: boolean;
+
+import hyphenate from "./hyphenate-style-name.js";
+import {prefix} from "inline-style-prefixer";
+
+import type {StyleObject} from "styletron-standard";
+
+export default function injectStylePrefixed(
+  styles: StyleObject,
+  className: string,
+  globalPrefix: string,
+) {
+  const outputRules = [];
+  let plainBlock = "";
+  for (const originalKey in styles) {
+    const originalVal = styles[originalKey];
+    if (originalVal === void 0 || originalVal === null) {
+      continue;
+    }
+    if (typeof originalVal !== "object") {
+      // Non-null and non-undefined primitive value
+      if (__DEV__) {
+        validateValueType(originalVal, originalKey);
+      }
+      const propValPair = `${hyphenate(
+        originalKey,
+      )}:${((originalVal: any): string)}`;
+      const prefixed = prefix({[originalKey]: originalVal});
+      for (const prefixedKey in prefixed) {
+        const prefixedVal = prefixed[prefixedKey];
+        const prefixedValType = typeof prefixedVal;
+        if (prefixedValType === "string" || prefixedValType === "number") {
+          const prefixedPair = `${hyphenate(prefixedKey)}:${prefixedVal}`;
+          if (prefixedPair !== propValPair) {
+            plainBlock += `${prefixedPair};`;
+          }
+        } else if (Array.isArray(prefixedVal)) {
+          const hyphenated = hyphenate(prefixedKey);
+          for (let i = 0; i < prefixedVal.length; i++) {
+            const prefixedPair = `${hyphenated}:${prefixedVal[i]}`;
+            if (prefixedPair !== propValPair) {
+              plainBlock += `${prefixedPair};`;
+            }
+          }
+        }
+      }
+      plainBlock += `${propValPair};`;
+    } else if (originalKey[0] === ":") {
+      outputRules.push(
+        `.${globalPrefix}css-${className}${originalKey}{${injectStylePrefixed(
+          originalVal,
+          "",
+          globalPrefix,
+        ).join("")}}`,
+      );
+    } else if (originalKey[0] === "@") {
```

Comment:
I think we should consider aligning the `@media` ordering strategy between the atomic and monolithic engine so that style object semantics remain as close as possible.

It seems like there's two options:
1. Add sorting to the monolithic engine. As we discussed though, this is less flexible for developers.
2. Replace sorting in the atomic engine with iteration order-based precedence. This could potentially be an option that is disabled by default to avoid a breaking change.

Reply:
In that case, I think number 2 makes sense:

1. We don't need to add this arbitrary limitation into monolithic engine
2. It will make the atomic engine more flexible
3. The implementation could share the logic with planning `@supports` support